### PR TITLE
Feature/mqtt agent publish

### PR DESCRIPTION
**Pull request by SgtSilvio**

### CODE REVIEW

File: src/main/java/org/varroa/chunk/command/generation/TopicNameGenerators.java [hunk 1]
```diff
@@ -32,8 +33,8 @@
     private final @NotNull Cache<String, TopicNameGenerator> generators;
     private final @NotNull ConcurrentHashMap<OffsetKey, AtomicInteger> offsets;
 
+    @Inject
```
Comment (sauroter) [inline src/main/java/org/varroa/chunk/command/generation/TopicNameGenerators.java]: why?

File: src/main/java/org/varroa/chunk/command/generation/TopicNameGenerators.java [hunk 2]
```diff
@@ -47,7 +48,7 @@ public TopicNameGenerators() {
         final int strippedOffset = (mqttAgentOffset >> trailingZeros) << trailingZeros;
 
         final AtomicInteger currentOffset =
-                offsets.computeIfAbsent(new OffsetKey(pattern, strippedOffset), NEW_OFFSET_FUNCTION);
+                offsets.computeIfAbsent(new OffsetKey(pattern, strippedOffset % topicGroupSize), NEW_OFFSET_FUNCTION);
```
Comment (sauroter) [inline src/main/java/org/varroa/chunk/command/generation/TopicNameGenerators.java]: imho this is wrong, as than more clients are mapped to the same key. The modulo should be applied in the offset function.

File: src/main/java/org/varroa/load/Mqtt5Agent.java [hunk 3]
```diff
@@ -58,7 +66,35 @@
...
+            rateFlowable = rateFlowable.zipWith(Flowable.interval(rate, TimeUnit.NANOSECONDS).onBackpressureDrop(),
+                    (offset, ignored) -> offset, false, (int) Math.min(count, 64));
+        }
+
+        final Flowable<Mqtt5PublishResult> publishResultFlowable;
+
+        if (publishCommand.isWaitForAck()) {
+            publishResultFlowable =
+                    rateFlowable.flatMap(
+                            offset -> client.publish(Flowable.just(getPublish(topicNameGeneration, qos))), 1);
```
Comment (sauroter) [inline src/main/java/org/varroa/load/Mqtt5Agent.java]: please add comments, for people who are not fluent in the fluent API ;)

File: src/main/java/org/varroa/load/MqttAgentGroup.java [hunk 4]
```diff
@@ -69,7 +69,7 @@ public MqttAgentGroup(final @NotNull MqttAgentDependencies dependencies, final @
             final MqttAgent mqttAgent = subPool[chunkPosition % MqttAgentPool.SUB_SIZE];
             lifecycles[chunkPosition] = mqttAgent.execute(chunkMessage.getChunk());
         }
-        return Completable.mergeArray(lifecycles);
+        return Completable.mergeArrayDelayError(lifecycles);
```
Comment (sauroter) [inline src/main/java/org/varroa/load/MqttAgentGroup.java]:
```suggestion
        //TODO: error handling
        return Completable.mergeArrayDelayError(lifecycles);
```

File: src/main/java/org/varroa/chunk/command/generation/TopicNameGenerators.java (hunk 1 above)
Comment (SgtSilvio) [reply src/main/java/org/varroa/chunk/command/generation/TopicNameGenerators.java]: IntelliJ reported that this constructur is not used. With the annotation it recognizes that it is created by injection

Comment (sauroter) [reply src/main/java/org/varroa/chunk/command/generation/TopicNameGenerators.java]: :(

File: src/main/java/org/varroa/chunk/command/generation/TopicNameGenerators.java (hunk 2 above)
Comment (SgtSilvio) [reply src/main/java/org/varroa/chunk/command/generation/TopicNameGenerators.java]: fixed

File: src/main/java/org/varroa/load/MqttAgentGroup.java (hunk 4 above)
Comment (SgtSilvio) [reply src/main/java/org/varroa/load/MqttAgentGroup.java]: I don't think that we should handle errors here, but either in the MqttAgent or the AgentHandshakeHandler

Comment (sauroter) [reply src/main/java/org/varroa/load/MqttAgentGroup.java]: yeah but we must do it somewhere

File: src/main/java/org/varroa/load/Mqtt5Agent.java [hunk 5]
```diff
@@ -74,17 +74,20 @@
 
         Flowable<Long> rateFlowable = Flowable.rangeLong(0, count);
         if (rate != TimeParser.NO_RATE) {
+            // if the interval is exceeded we buffer up to 32 ticks and then drop the following
```
Comment (sauroter) [inline src/main/java/org/varroa/load/Mqtt5Agent.java]: 👍

**Pull request merged**